upgrade/util/util.go

```diff
@@ -149,6 +160,77 @@ func CreateOrUpdateLonghornVersionSetting(namespace string, lhClient *lhclientse
 	return nil
 }
 
+// IsUpgradePathSupported returns if the upgrade path from lhVersionBeforeUpgrade to meta.Version is supported.
+func IsUpgradePathSupported(namespace string, lhClient *lhclientset.Clientset) bool {
+	lhVersionBeforeUpgrade, err := GetCurrentLonghornVersion(namespace, lhClient)
+	if err != nil {
+		logrus.WithError(err).Warn("failed to get current longhorn version")
+		return false
+	}
+
+	if lhVersionBeforeUpgrade == "" {
+		return true
+	}
+
+	logrus.Infof("Checking if the upgrade path (%v to %v) is supported", lhVersionBeforeUpgrade, meta.Version)
+
+	if !semver.IsValid(meta.Version) {
+		logrus.Warnf("Upgrading version is not valid: %v", meta.Version)
+		return false
+	}
+
+	lhNewMajorVersion := semver.Major(meta.Version)
+	lhMajorVersionBeforeUpgrade := semver.Major(lhVersionBeforeUpgrade)
+
+	lhNewMinorVersionNum, err := getMinorInt(meta.Version)
+	if err != nil {
+		logrus.WithError(err).Warnf("failed to parse %v minor version", meta.Version)
+		return false
+	}
+
+	lhMinorVersionBeforeUpgradeNum, err := getMinorInt(lhVersionBeforeUpgrade)
+	if err != nil {
+		logrus.WithError(err).Warnf("failed to parse %v minor version", lhVersionBeforeUpgrade)
+		return false
+	}
+
+	if semver.Compare(lhMajorVersionBeforeUpgrade, lhNewMajorVersion) > 0 {
+		logrus.Warnf("Downgrading is not supported, %v to %v", lhVersionBeforeUpgrade, meta.Version)
+		return false
+	}
+
+	if semver.Compare(lhMajorVersionBeforeUpgrade, lhNewMajorVersion) < 0 {
+		return lhMinorVersionBeforeUpgradeNum >= LonghornV1ToV2MinorVersionNum
+	}
+
+	if (lhNewMinorVersionNum-lhMinorVersionBeforeUpgradeNum) > 1 || (lhNewMinorVersionNum-lhMinorVersionBeforeUpgradeNum) < 0 {
+		return false
+	}
+
+	if (lhNewMinorVersionNum - lhMinorVersionBeforeUpgradeNum) == 1 {
+		return true
+	}
+
+	if semver.Compare(lhVersionBeforeUpgrade, meta.Version) > 0 {
+		logrus.Warnf("Downgrading is not supported, %v to %v", lhVersionBeforeUpgrade, meta.Version)
+		return false
+	}
```

Comment:
Is it better to check the patch version then rather than directly comparing the whole version number? Will this error out when the version is from `v1.4.0-rc1` to `v.1.4.0-rc2`?

Reply:
Upgrading from `v1.4.0-rc1` to `v.1.4.0-rc2` will be allowed, but `v1.4.0-rc2` to `v.1.4.0-rc1` will not.
`semver.Compare` can check the `prerelease` version.
Do we allow downgrade for pre-release version?

Reply:
I am fine with either... If we don't need this kind of pre-release version check, we can check the patch version only instead, right?

Reply:
Yes, I think we could do that instead.